### ISSUE TITLE
ExponentialSmoothing: AssertionError  ub > lb

### ISSUE THREAD

**Issue by rschwarz** [LAMPSPUC/StateSpaceModels.jl#287]
I'm trying to fit an exponential smoothing model on (integer-valued) time series data, but I'm getting an error during the `fit!` method.

Here's the code snippet:
```julia
y = Float64.(by_month[!, :Count])
model = ExponentialSmoothing(y, trend=true, seasonal=12)
fit!(model)
```
and the stack trace
```julia
AssertionError: ub > lb

Stacktrace:
  [1] constrain_box!(model::ExponentialSmoothing, str::String, lb::Float64, ub::Float64)
    @ StateSpaceModels ~/.julia/packages/StateSpaceModels/JApj1/src/hyperparameters.jl:327
  [2] constrain_hyperparameters!(model::ExponentialSmoothing)
    @ StateSpaceModels ~/.julia/packages/StateSpaceModels/JApj1/src/models/exponential_smoothing.jl:177
  [3] update_model_hyperparameters!
    @ ~/.julia/packages/StateSpaceModels/JApj1/src/kalman_filter_and_smoother.jl:22 [inlined]
  [4] optim_loglike(model::ExponentialSmoothing, filter::UnivariateKalmanFilter{Float64}, unconstrained_hyperparameters::Vector{Float64})
    @ StateSpaceModels ~/.julia/packages/StateSpaceModels/JApj1/src/kalman_filter_and_smoother.jl:11
  [5] #41
    @ ~/.julia/packages/StateSpaceModels/JApj1/src/fit.jl:38 [inlined]
  [6] finite_difference_gradient!(df::Vector{Float64}, f::StateSpaceModels.var"#41#42"{UnivariateKalmanFilter{Float64}, ExponentialSmoothing}, x::Vector{Float64}, cache::FiniteDiff.GradientCache{Nothing, Nothing, Nothing, Vector{Float64}, Val{:central}(), Float64, Val{true}()}; relstep::Float64, absstep::Float64, dir::Bool)
    @ FiniteDiff ~/.julia/packages/FiniteDiff/blirf/src/gradients.jl:273
  [7] finite_difference_gradient!(df::Vector{Float64}, f::Function, x::Vector{Float64}, cache::FiniteDiff.GradientCache{Nothing, Nothing, Nothing, Vector{Float64}, Val{:central}(), Float64, Val{true}()})
    @ FiniteDiff ~/.julia/packages/FiniteDiff/blirf/src/gradients.jl:224
  [8] (::NLSolversBase.var"#g!#44"{StateSpaceModels.var"#41#42"{UnivariateKalmanFilter{Float64}, ExponentialSmoothing}, FiniteDiff.GradientCache{Nothing, Nothing, Nothing, Vector{Float64}, Val{:central}(), Float64, Val{true}()}})(storage::Vector{Float64}, x::Vector{Float64})
    @ NLSolversBase ~/.julia/packages/NLSolversBase/geyh3/src/objective_types/twicedifferentiable.jl:113
  [9] (::NLSolversBase.var"#fg!#45"{StateSpaceModels.var"#41#42"{UnivariateKalmanFilter{Float64}, ExponentialSmoothing}})(storage::Vector{Float64}, x::Vector{Float64})
    @ NLSolversBase ~/.julia/packages/NLSolversBase/geyh3/src/objective_types/twicedifferentiable.jl:117
 [10] value_gradient!!(obj::NLSolversBase.TwiceDifferentiable{Float64, Vector{Float64}, Matrix{Float64}, Vector{Float64}}, x::Vector{Float64})
    @ NLSolversBase ~/.julia/packages/NLSolversBase/geyh3/src/interface.jl:82
 [11] value_gradient!(obj::NLSolversBase.TwiceDifferentiable{Float64, Vector{Float64}, Matrix{Float64}, Vector{Float64}}, x::Vector{Float64})
    @ NLSolversBase ~/.julia/packages/NLSolversBase/geyh3/src/interface.jl:69
 [12] value_gradient!(obj::Optim.ManifoldObjective{NLSolversBase.TwiceDifferentiable{Float64, Vector{Float64}, Matrix{Float64}, Vector{Float64}}}, x::Vector{Float64})
    @ Optim ~/.julia/packages/Optim/uwNqi/src/Manifolds.jl:50
 [13] (::LineSearches.var"#ϕdϕ#6"{Optim.ManifoldObjective{NLSolversBase.TwiceDifferentiable{Float64, Vector{Float64}, Matrix{Float64}, Vector{Float64}}}, Vector{Float64}, Vector{Float64}, Vector{Float64}})(α::Float64)
    @ LineSearches ~/.julia/packages/LineSearches/Ki4c5/src/LineSearches.jl:84
 [14] (::LineSearches.HagerZhang{Float64, Base.RefValue{Bool}})(ϕ::Function, ϕdϕ::LineSearches.var"#ϕdϕ#6"{Optim.ManifoldObjective{NLSolversBase.TwiceDifferentiable{Float64, Vector{Float64}, Matrix{Float64}, Vector{Float64}}}, Vector{Float64}, Vector{Float64}, Vector{Float64}}, c::Float64, phi_0::Float64, dphi_0::Float64)
    @ LineSearches ~/.julia/packages/LineSearches/Ki4c5/src/hagerzhang.jl:139
 [15] HagerZhang
    @ ~/.julia/packages/LineSearches/Ki4c5/src/hagerzhang.jl:101 [inlined]
 [16] perform_linesearch!(state::Optim.LBFGSState{Vector{Float64}, Vector{Vector{Float64}}, Vector{Vector{Float64}}, Float64, Vector{Float64}}, method::Optim.LBFGS{Nothing, LineSearches.InitialStatic{Float64}, LineSearches.HagerZhang{Float64, Base.RefValue{Bool}}, Optim.var"#17#19"}, d::Optim.ManifoldObjective{NLSolversBase.TwiceDifferentiable{Float64, Vector{Float64}, Matrix{Float64}, Vector{Float64}}})
    @ Optim ~/.julia/packages/Optim/uwNqi/src/utilities/perform_linesearch.jl:59
 [17] update_state!(d::NLSolversBase.TwiceDifferentiable{Float64, Vector{Float64}, Matrix{Float64}, Vector{Float64}}, state::Optim.LBFGSState{Vector{Float64}, Vector{Vector{Float64}}, Vector{Vector{Float64}}, Float64, Vector{Float64}}, method::Optim.LBFGS{Nothing, LineSearches.InitialStatic{Float64}, LineSearches.HagerZhang{Float64, Base.RefValue{Bool}}, Optim.var"#17#19"})
    @ Optim ~/.julia/packages/Optim/uwNqi/src/multivariate/solvers/first_order/l_bfgs.jl:204
 [18] optimize(d::NLSolversBase.TwiceDifferentiable{Float64, Vector{Float64}, Matrix{Float64}, Vector{Float64}}, initial_x::Vector{Float64}, method::Optim.LBFGS{Nothing, LineSearches.InitialStatic{Float64}, LineSearches.HagerZhang{Float64, Base.RefValue{Bool}}, Optim.var"#17#19"}, options::Optim.Options{Float64, Nothing}, state::Optim.LBFGSState{Vector{Float64}, Vector{Vector{Float64}}, Vector{Vector{Float64}}, Float64, Vector{Float64}})
    @ Optim ~/.julia/packages/Optim/uwNqi/src/multivariate/optimize/optimize.jl:57
 [19] optimize(d::NLSolversBase.TwiceDifferentiable{Float64, Vector{Float64}, Matrix{Float64}, Vector{Float64}}, initial_x::Vector{Float64}, method::Optim.LBFGS{Nothing, LineSearches.InitialStatic{Float64}, LineSearches.HagerZhang{Float64, Base.RefValue{Bool}}, Optim.var"#17#19"}, options::Optim.Options{Float64, Nothing})
    @ Optim ~/.julia/packages/Optim/uwNqi/src/multivariate/optimize/optimize.jl:35
 [20] fit!(model::ExponentialSmoothing; filter::UnivariateKalmanFilter{Float64}, optimizer::Optimizer)
    @ StateSpaceModels ~/.julia/packages/StateSpaceModels/JApj1/src/fit.jl:40
 [21] fit!(model::ExponentialSmoothing)
    @ StateSpaceModels ~/.julia/packages/StateSpaceModels/JApj1/src/fit.jl:33
 [22] top-level scope
    @ In[24]:3
 [23] eval
    @ ./boot.jl:360 [inlined]
 [24] include_string(mapexpr::typeof(REPL.softscope), mod::Module, code::String, filename::String)
    @ Base ./loading.jl:1094
```

This is on Julia 1.6.0, with StateSpaceModels at 0.5.17.

Unfortunately, I can't share the data to reproduce now, but I'll try to trigger it with random data instead for an MWE.

**Comment by rschwarz** [comment]
The following example also triggers the same error:
```julia
using StateSpaceModels

# some years of monthly data
years = 4
month = collect(1:(years * 12))

# deterministic data
start = 600
season = 100 .* sin.((2π/12) .* month)
trend = -month
noise = month .% 5

# integer values, as Float64
y = ceil.(start .+ trend .+ season .+ noise)

# fit model with trend and season
model = ExponentialSmoothing(y, trend=true, seasonal=12)
fit!(model)
```

Within the `@assert`, we have `lb == ub == 0.0001`. So I suspect that the [initialization](https://github.com/LAMPSPUC/StateSpaceModels.jl/blob/master/src/models/exponential_smoothing.jl#L176) could be to blame. Should it be `max` instead of `min` here?

**Comment by guilhermebodin** [comment]
Hi @rschwarz thank you for pointing the issue.

I made a branch fixing the bug, the fact is that the upper bound limit for the seasonal smoothing hyperparameter was not inside the recommended bounds.

Please add the package on branch `fix-bug-es` by doing `add StateSpaceModels#fix-bug-es` and give it a try on your data. :)

I would also recommend trying the `UnobservedComponents model`. 
```julia
model = UnobservedComponents(y, trend = "local linear trend", seasonal="stochastic 12")
````

**Comment by rschwarz** [comment]
Thanks, I tried your branch and no longer got the AssertionError.
Further, the forecast result look better (compared to my `min -> max` "fix" which left some parameters at 0.0001).

**Comment by guilhermebodin** [comment]
Nice! I am going to make a release and close the issue.
